src/realm/sync/binding_callback_thread_observer.hpp

```diff
@@ -20,23 +20,125 @@
 #define REALM_OS_BINDING_CALLBACK_THREAD_OBSERVER_HPP
 
 #include <exception>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <thread>
+
 
 namespace realm {
 // Interface for bindings interested in registering callbacks before/after the ObjectStore thread runs.
 // This is for example helpful to attach/detach the pthread to the JavaVM in order to be able to perform JNI calls.
-class BindingCallbackThreadObserver {
-public:
+struct BindingCallbackThreadObserver {
+    using NotificationCallback = std::function<void()>;
+    using ErrorCallback = std::function<void(const std::exception&)>;
+
+    // Create a BindingCallbackThreadObserver that can be used in SyncClientConfig
+    BindingCallbackThreadObserver(std::optional<NotificationCallback>&& did_create_thread,
+                                  std::optional<NotificationCallback>&& will_destroy_thread,
+                                  std::optional<ErrorCallback>&& error_handler)
+        : m_create_thread_callback{std::move(did_create_thread)}
+        , m_destroy_thread_callback{std::move(will_destroy_thread)}
+        , m_handle_error_callback{std::move(error_handler)}
+    {
+    }
+
+    virtual ~BindingCallbackThreadObserver() = default;
+
+    // Set the global thread observer with the provided (optional) callback functions
+    static void set_global_thread_observer(std::unique_ptr<BindingCallbackThreadObserver>&& observer_ptr)
```

Comment:
Do we really need to keep the old global API here? Currently this is used by just one SDK. @cmelchior, is adopting the new per-app API sufficiently complicated for java that we should support both APIs for now?

Reply:
No, it shouldn't be complicated. Removing this is fine 👍

Reply:
OK - I'll remove the global support and update the CAPI.